### PR TITLE
Add intel-oneapi-compiler into pcluster image

### PR DESCRIPTION
**Pull request by stephenmsachs**
- This image is intended only to be used within the spack gitlab runners.
- Remove as much as possible after installation to minimize image size.

### CODE REVIEW

**Comment by stephenmsachs** [comment]
This supersedes  @rscohn2's https://github.com/spack/gitlab-runners/pull/51.

**Review by rscohn2 — APPROVED** [review]
Looks good. Thanks!

**Comment by stephenmsachs** [comment]
@alalazo can you please review? I ran https://gitlab.spack.io/spack/spack/-/pipelines/703987 with this container. It seems to work as expected.

One thing I am not sure of is the `spack clean -a`. I want to minimize the image size, but at the same time not waste time in the buildcache stack. I see there is some time spent in `source spack/share/spack/setup-env.sh` in the stack containers. Are there temporary files deleted during `spack clean -a` which get re-created during `source spack/share/spack/setup-env.sh`? If this is the case, we could use a finer granularity `spack clean`.

**Review by alalazo — CHANGES_REQUESTED** [review]
I inspected the last layer, not sure the modifications are what we expect. Can you comment / confirm:

Screenshot from 2024-05-27 16-12-30

I checked the layer using https://github.com/wagoodman/dive

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile [hunk 1]
```diff
@@ -141,6 +141,30 @@ RUN --mount=type=secret,id=bootstrap_gcc_key \
...
+    && cp "${SPACK_ROOT}/share/spack/gitlab/cloud_pipelines/configs/config.yaml" "/bootstrap/cloud_pipelines-config.yaml" \
+    && . "${SPACK_ROOT}/share/spack/setup-env.sh" \
+    && spack load gcc \
+    && spack compiler add --scope site \
+    && spack external find --scope site --tag core-packages \
+    && cd "${SPACK_ROOT}" \
+    && curl -sL https://github.com/spack/spack/pull/40557.patch | patch -p1 \
+    && curl -sL https://github.com/spack/spack/pull/40561.patch | patch -p1 \
+    && spack install intel-oneapi-compilers-classic \
+    && rm -rf $(find $(spack find --format '{prefix}' intel-oneapi-compilers) -type d -name 'conda_channel') \
```
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:160]: The classic compilers are just a wrapper around the actual `intel-oneapi-compilers`. I clean up in the actual compilers directories to free up some space where possible.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile [hunk 2]
```diff
@@ -141,6 +141,30 @@ RUN --mount=type=secret,id=bootstrap_gcc_key \
     && spack gpg publish --rebuild-index -m bootstrap-gcc-cache ${secretkey_fingerprint} \
     && rm -rf $(dirname "${SPACK_ROOT}") /root/.spack
 
+# Add intel compiler by using a constant git tag. Leave the spack installation to enable
+# `spack install intel-oneapi-compiler` during container bootstrap.
+ARG SPACK_TAG="develop-2023-08-06"
```
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:146]: When a gitlab stack is running using this container, the concretizer will identify some of the dependencies, e.g. `cmake@3.27.9%intel@2021.10.0` as pre-existing in the binary cache. This `cmake` includes some `RPATH` entries to C++ libraries inside the Intel compiler's directory. So, the `intel-oneapi-compilers` package needs to have the same location (e.g. the same hash) as it had when we first started populating the cache with Intel compiled packages.
We can update to a newer `SPACK_TAG` as soon as we use a newer Intel compiler version. Unfortunately this is currently the last one with full classic [C compiler support](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/intel-oneapi-compilers-classic/package.py#L61).

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile [hunk 3]
```diff
@@ -141,6 +141,30 @@ RUN --mount=type=secret,id=bootstrap_gcc_key \
...
+    && git clone --depth=1 -b ${SPACK_TAG} https://github.com/spack/spack.git "${SPACK_ROOT}" \
+    && cp "${SPACK_ROOT}/share/spack/gitlab/cloud_pipelines/configs/config.yaml" "${SPACK_ROOT}/etc/spack/config.yaml" \
+    && cp "${SPACK_ROOT}/share/spack/gitlab/cloud_pipelines/configs/config.yaml" "/bootstrap/cloud_pipelines-config.yaml" \
+    && . "${SPACK_ROOT}/share/spack/setup-env.sh" \
+    && spack load gcc \
+    && spack compiler add --scope site \
+    && spack external find --scope site --tag core-packages \
+    && cd "${SPACK_ROOT}" \
+    && curl -sL https://github.com/spack/spack/pull/40557.patch | patch -p1 \
+    && curl -sL https://github.com/spack/spack/pull/40561.patch | patch -p1 \
```
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:158]: The same reason as above. These patches are necessary to make the Fortran compiler work, but had not been merged in the last available github tag when we added the intel based stacks.

**Comment by stephenmsachs** [comment]
> I inspected the last layer, not sure the modifications are what we expect. Can you comment / confirm:
> 
> Screenshot from 2024-05-27 16-12-30
> 
> I checked the layer using https://github.com/wagoodman/dive

Looks good to me, except for the `/etc/OpenCL`. I don't believe we need this. @rscohn2 do you know if we can delete those files?

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:160]: I'm confused, which compilers would you need? The classic compilers have been deprecated, and currently only `ifort` is shipped with the last `oneapi` releases.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:146]: Did you check the work done on `oneapi` and `intel-oneapi-runtime`? If that's the only reason, we'd better use `%oneapi` and rely on the runtime we extract.

**Comment by alalazo** [comment]
What do we use the old `spack` instance for? The one under `bootstrap-compilers/spack`. 

Also, @rscohn2 the `intel-oneapi` installation will be pushed into a public image on ghcr.io (see layer above), is that fine?

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (rscohn2) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:160]: The preferred version for intel-oneapi-compilers-classic is the latest release that has both icc and ifort.  I guess the issue is that this package has 2023 oneapi compilers, and 2024 and later compilers will not be tested. But that is something that could be improved over time.

**Comment by rscohn2** [comment]
> Looks good to me, except for the `/etc/OpenCL`. I don't believe we need this. @rscohn2 do you know if we can delete those files?

opencl is needed to support SYCL on CPU. The files are just the path of the shared objects so there is no space savings from deleting it.

**Comment by rscohn2** [comment]
> Also, @rscohn2 the `intel-oneapi` installation will be pushed into a public image on ghcr.io (see layer above), is that fine?

Its purpose is to support CI so it doesn't seem like redistribution to me. I cannot speak definitively for Intel. I believe this activity is moving under HPSF and we can look at all the related questions together. If the public image turns out not to be ok, I am sure we can find some other solution.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:160]: I think you are right, we can remove the the classic part altogether and pin the `intel-oneapi-compiler` version. Let me try that.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile:146]: Let me look into that.

[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile]: Using oneapi-runtime now.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile]: No need to use classic compiler. Moved to using the latest oneapi compiler compatible with Amazon Linux2 `glibc`.

**Comment by stephenmsachs** [comment]
I addressed the comments here and in https://github.com/spack/spack/pull/44150. It's ready to be reveiwed again.

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenmsachs) [reply Dockerfiles/pcluster-amazonlinux-2/Dockerfile]: Patches not necessary any more since we are using a newer spack tag.

**Review by alalazo — APPROVED** [review]

File: Dockerfiles/pcluster-amazonlinux-2/Dockerfile [hunk 4]
```diff
@@ -141,6 +136,28 @@ RUN --mount=type=secret,id=bootstrap_gcc_key \
...
+    && git clone --depth=1 -b ${SPACK_TAG} https://github.com/spack/spack.git "${SPACK_ROOT}" \
+    && cp "${SPACK_ROOT}/share/spack/gitlab/cloud_pipelines/configs/config.yaml" "${SPACK_ROOT}/etc/spack/config.yaml" \
+    && cp "${SPACK_ROOT}/share/spack/gitlab/cloud_pipelines/configs/config.yaml" "/bootstrap/cloud_pipelines-config.yaml" \
+    && . "${SPACK_ROOT}/share/spack/setup-env.sh" \
+    && spack load gcc \
+    && spack compiler add --scope site \
+    && cd "${SPACK_ROOT}" \
+    && spack install intel-oneapi-compilers@2024.1.0 \
+    && . "$(spack location -i intel-oneapi-compilers)"/setvars.sh \
+    && spack compiler add --scope site \
```
Comment (mabraham) [inline Dockerfiles/pcluster-amazonlinux-2/Dockerfile:152]: https://spack.readthedocs.io/en/latest/build_systems/inteloneapipackage.html suggests this line should be more like
```
spack compiler add `spack location -i intel-oneapi-compilers`/compiler/latest/bin --scope site \
```

**Pull request merged**